### PR TITLE
Adapt KHeader parameters

### PR DESCRIPTION
**Pull request by ckolbPTB**

### CODE REVIEW

File: src/mrpro/data/AcqInfo.py [hunk 1]
```diff
@@ -236,6 +219,23 @@ def spatialdimension_2d(
             user7=tensor(idx['user'][:, 7]),
         )
 
+        # Calculate orientation as rotation matrix from directional cosines
+        def orientation_from_directional_cosine(
```
Comment (ckolbPTB) [inline src/mrpro/data/AcqInfo.py]: Not sure if this is correct, please double check

File: src/mrpro/data/TrajectoryDescription.py [hunk 2]
```diff
@@ -1,29 +0,0 @@
-"""TrajectoryDescription dataclass."""
```
Comment (ckolbPTB) [inline src/mrpro/data/TrajectoryDescription.py]: I removed this because we are not using it anywhere and I do not see much use in supporting that

**Comment by ckolbPTB** [comment]
One problem: `acq_info.orientation: Rotation` is not modified by `modify_acq_info`. Adapting `modify_acq_info` to adapt `acq_info.orientation: Rotation` would lead to a circular import...

maybe easier to keep the orientation as read_dir, phase_dir and slice_dir?

**Comment by fzimmermann89** [comment]
Suggestion:
Rename cmodify_acq_info to ._apply
and move it to the acq info class.
It does pretty much the same as the _apply on a torch.module - apply a function to the tensor fields.
We could,to be more consistent,also name it _apply_ as it does in place modification.

Soon(tm) we will most likely need this kind of apply also on the other objects for the fancy indexing..

**Comment by fzimmermann89** [comment]
Alternatively (or additionally) the modify acq info should be changed (dependence injections):
Move the is dataclass condition up.
Do the other path on else (the previous tensor path)

Now, in the modify functions you can do a match case on what kind of object you get.
If it is something you do not want to change, do nothing.
For most modifications, rotation should behave the same as a tensor -  and you can just index into it in the modify function.

**Comment by fzimmermann89** [comment]
(Ps: sorry for the spelling, hope you can guess what I try to say. I can only communicate using  my phone and one hand at most, as my main job still seems to be a heating pad.. )

File: src/mrpro/data/_kdata/KDataSelectMixin.py [hunk 3]
```diff
@@ -49,10 +52,13 @@ def select_other_subset(
         other_idx = torch.cat([torch.where(idx == label_idx[:, 0, 0])[0] for idx in subset_idx], dim=0)
 
         # Adapt header
-        def select_acq_info(info: torch.Tensor):
-            return info[other_idx, ...]
+        def select_acq_info(field: T) -> T:
```
Comment (ckolbPTB) [inline src/mrpro/data/_kdata/KDataSelectMixin.py]: adapt after #417 merge

File: src/mrpro/data/_kdata/KDataSplitMixin.py [hunk 4]
```diff
@@ -94,10 +99,17 @@ def split_acq_info(acq_info: torch.Tensor) -> torch.Tensor:
         kheader = copy.deepcopy(self.header)
 
         # Update shape of acquisition info index
-        def reshape_acq_info(info: torch.Tensor):
-            return rearrange(split_acq_info(info), rearrange_pattern_acq_info)
-
-        kheader.acq_info = modify_acq_info(reshape_acq_info, kheader.acq_info)
+        def split_acq_info(field: T) -> T:
```
Comment (ckolbPTB) [inline src/mrpro/data/_kdata/KDataSplitMixin.py]: adapt after #417 merge

**Comment by ckolbPTB** [comment]
The _apply_ function works nicely. I moved it to DRAFT because we should first merge  #417 and then we can remove the special case for SpatialDimension in KDataSplitMixin and KDataSelectMixin

**Comment by ckolbPTB** [comment]
Something is not right with the orientation. 

Siemens saves the orientation as a quaternion. This is then translated to read_dir, slice_dir and phase_dir in siemens_to_ismrmrd here:
https://github.com/ismrmrd/siemens_to_ismrmrd/blob/cda30e2f4fbace05dce666a781e20a11eb28c7e6/main.cpp#L1139C14-L1139C46
using
https://github.com/ismrmrd/ismrmrd/blob/4037c41fcdcb8870234baaeb46ce4848aac1579a/libsrc/ismrmrd.c#L667

To me it seems that the signs are not necessarily correct or I am misunderstanding what these directions are supposed to represent.

File: src/mrpro/data/TrajectoryDescription.py (hunk 2 above)
Comment (fzimmermann89) [reply src/mrpro/data/TrajectoryDescription.py]: Isn't this one of the things 8.14 would use?
I thought they store something in these user values?

Comment (fzimmermann89) [reply src/mrpro/data/TrajectoryDescription.py]: I will ask max next week

File: src/mrpro/data/AcqInfo.py (hunk 1 above)
Comment (fzimmermann89) [reply src/mrpro/data/AcqInfo.py]: Isn't this transposed?

Applying the orientation to a 1,0,0 vector should give the slice_dir.
So slice dir hascto be the first column.

**Comment by fzimmermann89** [comment]
> Something is not right with the orientation. 
> 
> Siemens saves the orientation as a quaternion. This is then translated to read_dir, slice_dir and phase_dir in siemens_to_ismrmrd here:
> https://github.com/ismrmrd/siemens_to_ismrmrd/blob/cda30e2f4fbace05dce666a781e20a11eb28c7e6/main.cpp#L1139C14-L1139C46
> using
> https://github.com/ismrmrd/ismrmrd/blob/4037c41fcdcb8870234baaeb46ce4848aac1579a/libsrc/ismrmrd.c#L667
> 
> To me it seems that the signs are not necessarily correct or I am misunderstanding what these directions are supposed to represent. 

How and why did you notice? Where is the issue in your opinion? 

I faintly remember that there was something stupid going on either in ismrmd or in the Siemens convection regarding a sign of a single direction..

**Comment by ckolbPTB** [comment]
> How and why did you notice? Where is the issue in your opinion?

I noticed that when I do ismrmrd-slice-phase-readout-dir -> mrpro quaternion -> ismrmrd-slice-phase-readout-dir then I don't get the same result as I put in. 

Siemens is a quaternion which corresponds to a rotation matrix with columns phase - readout - slice encoding. If I take that into consideration (latest commit) I get the same out as I put in from the ismrmrd file. But I guess this still does not fit to our slice - phase - readout convention...

File: src/mrpro/data/TrajectoryDescription.py (hunk 2 above)
Comment (ckolbPTB) [reply src/mrpro/data/TrajectoryDescription.py]: any update from Max?

File: src/mrpro/data/AcqInfo.py (hunk 1 above)
Comment (ckolbPTB) [reply src/mrpro/data/AcqInfo.py]: should be fixed now

Comment (fzimmermann89) [reply src/mrpro/data/AcqInfo.py]: I think I added a function to rotation to create it from spatial dimensions...

File: src/mrpro/data/AcqInfo.py [hunk 5]
```diff
@@ -251,17 +270,37 @@ def spatialdimension_2d(
             flags=tensor_2d(headers['flags']),
             measurement_uid=tensor_2d(headers['measurement_uid']),
             number_of_samples=tensor_2d(headers['number_of_samples']),
+            orientation=orientation_from_directional_cosine(
+                spatialdimension_2d(headers['slice_dir']),
```
Comment (fzimmermann89) [inline src/mrpro/data/AcqInfo.py]: Suggestion: move the apply of conversion to here 
Such that SpatialDimension2d only creates a 2d SpatialDimension,  and the conversion is done explicitly in each line. Makes it easier to see what is happening

Comment (fzimmermann89) [reply src/mrpro/data/AcqInfo.py]: I can't comment on the SpatialDimension2d code where it should be changed..

File: src/mrpro/data/AcqInfo.py (hunk 1 above)
Comment (ckolbPTB) [reply src/mrpro/data/AcqInfo.py]: very handy

**Comment by fzimmermann89** [comment]
Maybe we can discuss UserParameters again I we make `misc` more private and unsupported

File: src/mrpro/data/AcqInfo.py [hunk 6]
```diff
@@ -251,17 +249,37 @@ def spatialdimension_2d(
...
             scan_counter=tensor_2d(headers['scan_counter']),
-            slice_dir=spatialdimension_2d(headers['slice_dir']),
             trajectory_dimensions=tensor_2d(headers['trajectory_dimensions']).fill_(3),  # see above
             user_float=tensor_2d(headers['user_float']),
             user_int=tensor_2d(headers['user_int']),
             version=tensor_2d(headers['version']),
         )
         return acq_info
+
+    def _apply_(self, modify_acq_info_field: Callable) -> None:
```
Comment (fzimmermann89) [inline src/mrpro/data/AcqInfo.py]: I think it would make sense it instead use the apply_ from the movedatamixin, as it should do the same, right?

Comment (ckolbPTB) [reply src/mrpro/data/AcqInfo.py]: yes, but I think we need #489 first, don't we?

Comment (fzimmermann89) [reply src/mrpro/data/AcqInfo.py]: Yes.. I wrote that in slack.. give me a bit of time to order things tonight and then merge them in a sane order

Comment (fzimmermann89) [reply src/mrpro/data/AcqInfo.py]: I am currently figuring out for me what a good order could be

**Comment by fzimmermann89** [comment]
I f'ed this up. sorry.

@ckolbPTB can you create a new PR from the unit_conversion_utils branch to main?

**Comment by fzimmermann89** [comment]
test

**Comment by fzimmermann89** [comment]
New PR: #506

**Pull request closed without merging**